### PR TITLE
224 add prisma to project

### PR DESCRIPTION
**Pull request by Elkrival**
Added prisma to project and updated some routes to get to home screen

this commit shows the updated routes to get past login screen 
https://github.com/AMP-SCZ/dpdash/pull/232/commits/ec59252a67978a296b86ec1e01bbd92898ab14bb

Screen Shot 2022-11-14 at 2 33 33 PM
Screen Shot 2022-11-14 at 2 34 16 PM

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -52,7 +53,7 @@
     "mathjs": "^7.6.0",
     "method-override": "^3.0.0",
     "moment": "^2.22.2",
-    "mongodb": "^3.6.6",
+    "mongodb": "^4.11.0",
```
Comment (mikestone14) [inline package.json:56]: is this change required to add prisma?

File: prisma/schema.prisma [hunk 2]
```diff
@@ -0,0 +1,122 @@
+generator client {
+  provider = "prisma-client-js"
+}
+
+datasource db {
+  provider = "mongodb"
+  url      = env("DPDMONGO_DATABASE_URL")
```
Comment (mikestone14) [inline prisma/schema.prisma:7]: what is `env`? Is this the same as `process.env`?

File: prisma/schema.prisma [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  /// Could not determine type: the field only had null or empty values in the sample set.
+  CA Json?
+  /// Could not determine type: the field only had null or empty values in the sample set.
+  LA Json?
+}
+
+model configs {
+  id      String        @id @default(auto()) @map("_id") @db.ObjectId
+  config  ConfigsConfig
+  created String
```
Comment (mikestone14) [inline prisma/schema.prisma:74]: is this a timestamp? If so, should it have an updated timestamp too?

File: prisma/schema.prisma [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  url      = env("DPDMONGO_DATABASE_URL")
+}
+
+type ConfigsConfig {
+  /// This field was commented out because of an invalid name. Please provide a valid one that matches [a-zA-Z][a-zA-Z0-9_]*
+  // 0 ConfigsConfig0[] @map("0")
+  colormap ConfigsConfigColormap[]
+}
+
+type ConfigsConfig0 {
```
Comment (mikestone14) [inline prisma/schema.prisma:16]: what is this type? Where is it used?

File: prisma/schema.prisma [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+  analysis String
+  category String
+  color    String[]
+  label    String
+  range    Int[]
+  text     Boolean
+  variable String
+}
+
+type ReportsCharts {
```
Comment (mikestone14) [inline prisma/schema.prisma:37]: this looks very similar to the charts we created. Are they intended to be the same?

File: prisma/schema.prisma [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+  reportName String
+  user       String @unique
+}
+
+model sessions {
+  id      String @id @default(auto()) @map("_id") @db.ObjectId
+  /// Field referred in an index, but found no data to define the type.
+  expires Json?
+
+  @@index([expires], map: "expires_1")
```
Comment (mikestone14) [inline prisma/schema.prisma:95]: what does the `map` input do here?

File: prisma/schema.prisma [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+model users {
+  id              String            @id @default(auto()) @map("_id") @db.ObjectId
+  access          String[]
+  /// Could not determine type: the field only had null or empty values in the sample set.
+  account_expires Json?
```
Comment (mikestone14) [inline prisma/schema.prisma:102]: can we remove this field if it is always empty or null?

File: prisma/schema.prisma [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+  complete UsersPreferencesComplete
+  config   String
+  sort     Int
+  star     UsersPreferencesStar
+}
+
+type UsersPreferencesComplete {
+  /// Could not determine type: the field only had null or empty values in the sample set.
+  CA Json?
+}
```
Comment (mikestone14) [inline prisma/schema.prisma:62]: do we need this if all the values are null or empty?

File: prisma/schema.prisma [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+type UsersPreferencesStar {
+  /// Could not determine type: the field only had null or empty values in the sample set.
+  CA Json?
+  /// Could not determine type: the field only had null or empty values in the sample set.
+  LA Json?
+}
```
Comment (mikestone14) [inline prisma/schema.prisma:69]: do we need this if all the values are null or empty?

File: prisma/schema.prisma [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+  variable String
+}
+
+type ReportsCharts {
+  assessmentSingle String
+  chartType        String
+  title            String
+  valueLabels      ReportsChartsValueLabels[]
+  /// Could not determine type: the field only had null or empty values in the sample set.
+  variableMulti    Json?
```
Comment (mikestone14) [inline prisma/schema.prisma:43]: do we need this if all the values are null or empty?

File: prisma/schema.prisma [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+  created    String
+  readers    String[]
+  reportName String
+  user       String @unique
+}
+
+model sessions {
+  id      String @id @default(auto()) @map("_id") @db.ObjectId
+  /// Field referred in an index, but found no data to define the type.
+  expires Json?
```
Comment (mikestone14) [inline prisma/schema.prisma:93]: does this comment mean that the field can be removed?

File: server/utils/passport/ensure-authenticated.js [hunk 12]
```diff
@@ -9,25 +9,23 @@ export default function ensureAuthenticated(req, res, next) {
...
-        switch(true) {
-          case !data || Object.keys(data).length === 0 || !!data.blocked || err:
-            return res.redirect(logoutForbiddenRoute)
-          case data.access && data.access.length === 0:
-            return res.redirect(unauthorizedRoute)
-          default:
-            return next()
-        }
+    })
+    .then((data) => {
```
Comment (mikestone14) [inline server/utils/passport/ensure-authenticated.js]: should we use async await here? Looks like we're converting the other promises to use this syntax.

File: server/routes/index.js [hunk 13]
```diff
@@ -779,120 +765,84 @@ router
...
-            console.log(err)
-            return res.status(502).send({ message: 'fail' })
-          } else {
-            return res.status(201).send({ message: 'success' })
-          }
+    try {
+      if (Object.prototype.hasOwnProperty.call(req.body, 'disable')) {
+        await appDb.configs.update({
+          where: { _id: new ObjectId(req.body.disable) },
+          data: { PULL: { readers: req.params.uid } },
```
Comment (mikestone14) [inline server/routes/index.js:774]: when is `PULL` used?

File: server/routes/index.js [hunk 14]
```diff
@@ -779,120 +765,84 @@ router
...
-        .collection('configs')
-        .deleteOne({ _id: new ObjectID(req.body.remove) }, function (err) {
-          if (err) {
-            console.log(err)
-            return res.status(502).send({ message: 'fail' })
-          } else {
-            return res.status(201).send({ message: 'success' })
-          }
+    try {
+      if (Object.prototype.hasOwnProperty.call(req.body, 'disable')) {
```
Comment (mikestone14) [inline server/routes/index.js:771]: I think we should create a ticket to refactor this function to make it more restful. IMO each of these `if/else` statements should be its own endpoint.

File: server/routes/index.js [hunk 15]
```diff
@@ -1399,13 +1350,15 @@ router
     try {
       const { appDb } = req.app.locals
       const { user } = req
-      const report = await appDb.collection('reports').findOne({
-        _id: ObjectID(req.params.id),
-        $or: [{ user }, { readers: user }],
+      const report = await appDb.reports.findUniqueOrThrow({
+        where: {
+          _id: ObjectId(req.params.id),
+          OR: [{ user }, { readers: user }],
```
Comment (mikestone14) [inline server/routes/index.js:1356]: does `{ readers: user }` mean that it will find reports where the user is included in the array of readers? We don't need to specify any sort of `IN` syntax to make that work?

File: server/routes/index.js [hunk 16]
```diff
@@ -1416,32 +1369,25 @@ router
...
-        },
-        {
-          $set: {
-            ...body,
-          },
-        }
-      )
+      await appDb.reports.update({
+        where: { user, _id: ObjectId(params.id) },
+        data: { ...body },
```
Comment (mikestone14) [inline server/routes/index.js:1374]: do we need to spread here?

File: prisma/schema.prisma (hunk 2 above)
Comment (Elkrival) [reply prisma/schema.prisma:7]: the env() is how prisma loads env variables similar to a node using process.env

File: prisma/schema.prisma (hunk 4 above)
Comment (Elkrival) [reply prisma/schema.prisma:16]: This type is generated from [prisma's introspection](https://www.prisma.io/docs/concepts/components/introspection). The correct typing will be visited on this ticket. 
Screen Shot 2022-11-14 at 4 17 56 PM

File: prisma/schema.prisma (hunk 5 above)
Comment (Elkrival) [reply prisma/schema.prisma:37]: There's a collection called reports in the db and it's separate than the charts functionality we worked on. I could ask during our sprint planning meeting tomorrow, I have zero context on how reports work.

File: prisma/schema.prisma (hunk 10 above)
Comment (Elkrival) [reply prisma/schema.prisma:43]: I'm unsure about how reports work, we also took them off the sidebar menu. I can ask the mgh folks about their plans on reports

File: prisma/schema.prisma (hunk 9 above)
Comment (Elkrival) [reply prisma/schema.prisma:69]: in the current state of the app we need this as this is how a user sets their preferred subjects. With the new schema we have an opportunity to change the type to something like.

type UsersStarSubject {
site: String,
starSubjects: String[]
}

Screen Shot 2022-11-14 at 4 28 36 PM

There's a ticket for the user schema. 
Screen Shot 2022-11-14 at 4 31 06 PM

File: prisma/schema.prisma (hunk 3 above)
Comment (Elkrival) [reply prisma/schema.prisma:74]: Thank you i'll add this to the config model ticket

File: prisma/schema.prisma (hunk 11 above)
Comment (Elkrival) [reply prisma/schema.prisma:93]: When I used introspection the app drops the collection.
https://github.com/AMP-SCZ/dpdash/blob/44f41f8e2c424efea78352e65da5e3edf25d0e48/server/app.js#L116

I'm sure that this is why it found no data. I'm adding a ticket for this schema. 

Screen Shot 2022-11-14 at 4 47 34 PM

File: prisma/schema.prisma (hunk 8 above)
Comment (Elkrival) [reply prisma/schema.prisma:62]: This would be a new schema that's the scope of another ticket.

File: prisma/schema.prisma (hunk 6 above)
Comment (Elkrival) [reply prisma/schema.prisma:95]: Not sure but there's a ticket for the sessions schema. 
Screen Shot 2022-11-14 at 4 49 25 PM

File: prisma/schema.prisma (hunk 7 above)
Comment (Elkrival) [reply prisma/schema.prisma:102]: I added it to the prisma user model ticket.
Screen Shot 2022-11-14 at 4 55 19 PM

**Review by Elkrival — COMMENTED** [review]
Coming along

File: server/utils/passport/ensure-authenticated.js (hunk 12 above)
Comment (Elkrival) [reply server/utils/passport/ensure-authenticated.js]: https://github.com/AMP-SCZ/dpdash/pull/232/commits/9c2ee658fa1d23719239cab98d359a8b7e8bf7d5

File: server/routes/index.js (hunk 14 above)
Comment (Elkrival) [reply server/routes/index.js:771]: Screen Shot 2022-11-14 at 5 07 15 PM
